File: back/kotlin/app/core/data/implementation/redis/GameRepositoryImpl.kt
package core.data.implementation.redis

class GameRepositoryImpl(private val redisData: RedisDataProtocol): GameRepository {
    
    override suspend fun createNewGame(newKey: String, board: StateType, player: PlayerType): Result<GameType> {
        return redisData.gameDao.setGame(newKey, board, player)
    }
    
    override suspend fun resetGame(gameID: String, board: StateType, player: PlayerType): Result<GameType> {
        return redisData.gameDao.resetGame(gameID, board, player)
    }
    
    override suspend fun updateBoard(gameID: String, position: CellPosition, player: PlayerType): Result<StateType>{
        return redisData.gameDao.addPlayerMove(gameID, Move(player, position))
    }
    
    override suspend fun getCurrentPlayer(gameID: String): Result<PlayerType> {
        return when (val result = redisData.gameDao.getInfo(gameID)) {
            is Result.Success -> {
                println("Data: ${result.data}")
                Result.Success(result.data.currentPlayer)
            }
            is Result.Error -> {
                println("Error: ${result.exception}")
                Result.Error(result.exception)
            }
            Result.NotFound -> Result.NotFound
        }
    }
    
    override suspend fun getBoardState(gameID: String): Result<StateType>{
        return when (val result = redisData.gameDao.getInfo(gameID)) {
                is Result.Success -> {
                    println("Data: ${result.data}")
                    Result.Success(result.data.state)
                }
                is Result.Error -> {
                    println("Error: ${result.exception}")
                    Result.Error(result.exception)
                }
                Result.NotFound -> Result.NotFound
            }
    }

    override suspend fun getGameState(gameID: String): Result<GameType>{
        return redisData.gameDao.getInfo(gameID)
    }

    override suspend fun updateGameState(gameID: String, board: StateType, info: GameInfo): Result<GameType> {
        return redisData.gameDao.updateInfo(gameID, board, info)
    }

}

File: back/go/app/core/data/implementation/redis/GameRepositoryImpl.go
package redis

import (
	"context"
	"go-ttt/app/core/database/redis/dao"
	"go-ttt/app/core/database/redis/entity"
	"go-ttt/app/core/model"
	"go-ttt/app/core/data/repository"
)

type GameRepositoryImpl struct {
	data dao.GameDao
}

func NewGameRepositoryImpl(Data dao.GameDao) repository.GameRepository {
	return &GameRepositoryImpl{data: Data}
}

func (r *GameRepositoryImpl) CreateNewGame(ctx context.Context, newKey string, board model.StateType, player model.PlayerType) (*model.GameType, error) {
	return r.data.SetGame(ctx, newKey, board, player)
}

func (r *GameRepositoryImpl) ResetGame(ctx context.Context, gameID string, board model.StateType, player model.PlayerType) (*model.GameType, error) {
	return r.data.ResetGame(ctx, gameID, board, player)
}

func (r *GameRepositoryImpl) UpdateBoard(ctx context.Context, gameID string, position model.CellPosition, player model.PlayerType) (*model.StateType, error) {
	playerMove := entity.Move{Player: player, Position: position}
	return r.data.AddPlayerMove(ctx, gameID, playerMove)
}

func (r *GameRepositoryImpl) GetCurrentPlayer(ctx context.Context, gameID string) (*model.PlayerType, error) {
	gameInfo, err := r.data.GetInfo(ctx, gameID)
	if err != nil {
		return nil, err
	}
	return &gameInfo.CurrentPlayer, nil
}

func (r *GameRepositoryImpl) GetBoardState(ctx context.Context, gameID string) (*model.StateType, error) {
	gameInfo, err := r.data.GetInfo(ctx, gameID)
		if err != nil {
			return nil, err
		}
		return &gameInfo.State, nil
}

func (r *GameRepositoryImpl) GetGameState(ctx context.Context, gameID string) (*model.GameType, error) {
	return r.data.GetInfo(ctx, gameID)
}

func (r *GameRepositoryImpl) UpdateGameState(ctx context.Context, gameID string, board model.StateType, info entity.GameInfo) (*model.GameType, error) {
	return r.data.UpdateInfo(ctx, gameID, board, info)
}